File: zig/space-age/space_age.zig
pub const Planet = enum {
    mercury,
    venus,
    earth,
    mars,
    jupiter,
    saturn,
    uranus,
    neptune,

    pub fn age(self: Planet, seconds: usize) f64 {
        const earth_years: f64 = @as(f64, @floatFromInt(seconds)) / 31_557_000.0;
        switch (self) {
            Planet.mercury => return earth_years / 0.2408467,
            Planet.venus => return earth_years / 0.61519726,
            Planet.earth => return earth_years / 1.0,
            Planet.mars => return earth_years / 1.8808158,
            Planet.jupiter => return earth_years / 11.862615,
            Planet.saturn => return earth_years / 29.447498,
            Planet.uranus => return earth_years / 84.016846,
            Planet.neptune => return earth_years / 164.79132,
        }
    }
};

File: elixir/space-age/lib/space_age.ex
defmodule SpaceAge do
  @type planet ::
          :mercury
          | :venus
          | :earth
          | :mars
          | :jupiter
          | :saturn
          | :uranus
          | :neptune

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet', or an error if 'planet' is not a planet.
  """
  @spec age_on(planet, pos_integer) :: {:ok, float} | {:error, String.t()}

  def age_on(:mercury, seconds), do: {:ok, seconds / 31_557_600 / 0.2408467}
  def age_on(:venus, seconds), do: {:ok, seconds / 31_557_600 / 0.61519726}
  def age_on(:earth, seconds), do: {:ok, seconds / 31_557_600}
  def age_on(:mars, seconds), do: {:ok, seconds / 31_557_600 / 1.8808158}
  def age_on(:jupiter, seconds), do: {:ok, seconds / 31_557_600 / 11.862615}
  def age_on(:saturn, seconds), do: {:ok, seconds / 31_557_600 / 29.447498}
  def age_on(:uranus, seconds), do: {:ok, seconds / 31_557_600 / 84.016846}
  def age_on(:neptune, seconds), do: {:ok, seconds / 31_557_600 / 164.79132}
  def age_on(planet, seconds), do: {:error, "not a planet"}

  # def age_on(planet, seconds) do
  #   case planet do
  #     :mercury -> {:ok, seconds / 31_557_600 / 0.2408467}
  #     :venus -> {:ok, seconds / 31_557_600 / 0.61519726}
  #     :earth -> {:ok, seconds / 31_557_600 / 1.0}
  #     :mars -> {:ok, seconds / 31_557_600 / 1.8808158}
  #     :jupiter -> {:ok, seconds / 31_557_600 / 11.862615}
  #     :saturn -> {:ok, seconds / 31_557_600 / 29.447498}
  #     :uranus -> {:ok, seconds / 31_557_600 / 84.016846}
  #     :neptune -> {:ok, seconds / 31_557_600 / 164.79132}
  #     _ -> {:error, "not a planet"}
  #   end
  # end
end
